searx/network/context.py

```diff
@@ -0,0 +1,387 @@
+        bound_instance = getattr(factory, "__self__", None)
+        if bound_instance is not None and hasattr(bound_instance, 'get_context'):
+            # bound_instance has a "get_context" attribute: this is most likely a Network
+            # searx.network.network.Network is not imported to avoid circular import
+            return f"<{common_attributes} network_context={factory.__self__!r}>"
+        # fallback : this instance was not created using Network.get_context
+        return f"<{common_attributes} http_client_factory={factory!r}>"
+
+
+## Measure time and deal with timeout
+
+
+class _TimeHTTPClientWrapper(ABCHTTPClient):
+    """Wrap an ABCHTTPClient:
+    * to record the HTTP runtime
+    * to override the timeout to make sure the total time does not exceed the timeout set on the NetworkContext
+    """
+
+    __slots__ = ('http_client', 'network_context')
+
+    @staticmethod
+    def wrap_factory(http_client_factory: HTTPCLIENTFACTORY, network_context: NetworkContext):
+        """Return a factory which wraps the result of http_client_factory with _TimeHTTPClientWrapper instance."""
+        functools.wraps(http_client_factory)
+
+        def wrapped_factory():
+            return _TimeHTTPClientWrapper(http_client_factory(), network_context)
+
+        wrapped_factory.__wrapped__ = http_client_factory
+        return wrapped_factory
+
+    def __init__(self, http_client: ABCHTTPClient, network_context: NetworkContext) -> None:
+        self.http_client = http_client
+        self.network_context = network_context
+
+    def send(self, stream, method, url, **kwargs) -> httpx.Response:
+        """Send the HTTP request using self.http_client
+
+        Inaccurate with stream: the method must record HTTP time with the close method of httpx.Response.
+        It is not a problem since stream are used only for the image proxy.
+        """
+        with self.network_context._record_http_time():  # pylint: disable=protected-access
+            timeout = self._extract_timeout(kwargs)
+            return self.http_client.send(stream, method, url, timeout=timeout, **kwargs)
+
+    def close(self):
+        return self.http_client.close()
+
+    @property
+    def is_closed(self) -> bool:
+        return self.http_client.is_closed
+
+    def _extract_timeout(self, kwargs):
+        """Extract the timeout parameter and adjust it according to the remaining time"""
+        timeout = kwargs.pop('timeout', None)
+        return self.network_context.get_remaining_time(timeout)
+
+
+## NetworkContextRetryFunction
+
+
+class NetworkContextRetryFunction(NetworkContext):
+    """When an HTTP request fails, this NetworkContext tries again
+    the whole function with another HTTP client.
+
+    This guarantees that func has the same HTTP client all along.
+    """
+
+    def call(self, func: Callable[P, R], *args: P.args, **kwargs: P.kwargs) -> R:
+        try:
+            while self._retries >= 0 and self.get_remaining_time() > 0:
```

Comment:
Even if retries is set to `0`, we still start one additional attempt. Is that intended?

Reply:
I don't think so.

Note: some tests are missing of the retry count.

Reply:
Fixed

Reply:
Actually: `retries` defines the number of ... retries, but at first the code must try.
Said differently, if for example `retries == 1`, the number of iteration must be `2`.
==> `self._retries >= 0` does not add an additional attempt.